rfcs/rdm-0079-collections.md

```diff
@@ -0,0 +1,337 @@
+
+![](0079/1.png)
+    
+### Technical implementation
+
+Naturally, collections are organized in tree-like structures. I.e. Cern Document Server (CDS) defines multiple top-level collections [1] that each have, potentially, other collections attached.
+    
+![1 - Cern Document Server collections example](0079/3.png)
+*[1] - CDS collections page, taken from https://cds.cern.ch/*
+    
+This can be represented in a tree (for simplicity only three levels are shown):
+    
+![](0079/4.png){width=1000}
+
+Another example of collections:
+    
+![](0079/5.png){width=500}
+*Euro Science subjects collections - http://publications.europa.eu/resource/dataset/euroscivoc*
+
+This solution proposes collections to be modelled as generic trees, whose information is persisted in a database. 
+    
+That said, collections data model will be based in the "Materialized Path" pattern. Read this [documentation from MongoDB](https://www.mongodb.com/docs/manual/tutorial/model-tree-structures-with-materialized-paths/) for a good overview of this pattern. The main concept is that each node (row in the DB) stores its `path` from the root to itself, excluding. For our example, it looks as follows:
+
+
+In our proposed model, however, a unique `id` is to be stored in the `path` so the ascendants of a node can be derived from the path and done with primary key lookups:
+    
+```jsonld
+   { _id: 1, title: "Natural Sciences", path: "," },
+   { _id: 2, title: "Physical science", path: ",1," },
+   { _id: 3, title: "Mathematics", path: ",1," },
+   { _id: 4, title: "Particles", path: ",1,2," },
+```
+
+**Database tables**
+
+We identified three new tables for this design:
+                                     
+1) `Collection`
+2) `CollectionTree`
+3) `CommunityCollectionTree`
+                                     
+![](0079/6.png)
+
+In our previous example, this is the final model and entities with the highlighted path of some collections (in red): 
+                                     
+![](0079/7.png)
+
+                                     
+### Design considerations 
+                                     
+The idea behind splitting `CollectionTree` from `Collection`, is to be able to define "root" collections that do not hold any query and simply group collections under a semantic umbrella. In our previous example, it would be modelled as such:
+                                     
+```
+European Open Repository  <- Community
+    Subjects                  <- CollectionTree
+        Natural sciences          <- Collection
+        Physical sciences         <- Collection
+```
+
+An important aspect of this design is that the `CommunityCollectionTree` allows the `community_id` to be set to `null`, indicating that the collection can be treated as a `root` collection. This enables the retrieval of all records from the repository without restricting them to a specific community.
```

Comment:
I am a bit lost on the null collection id. Since the communityCollectionTree table seems to be intermediary table between the collection tree and community, then can we simply skip creating an entry in CommunityCollectionTree when we define a tree with no collection? why to allow to set it to null ? unless this is not "relational" model in a standard meaning

Reply:
These are good questions, I will do my best to reply:

1. The table `CommunityCollectionTree` does not make sense anymore since we just want `A community has N collection_tree` and `A collection_tree belong to ONE community only`. So the table `CommunityCollectionTree` disappears - I will update the RFC afterwards.

2.  a `null` `community_id` means that the collection belongs to the whole repository and not only a community. An entry on `CollectionTree` is still needed, to define the `root` node of that tree.